### PR TITLE
fix(ds): enable filter on storybook and fix errors

### PR DESCRIPTION
**Pull request by murayama-r**
# Background

Unable to check datagrid filter on storybook
https://tailor.atlassian.net/browse/FL-146
onFilterChange handler should not be called if no filter change made
https://tailor.atlassian.net/browse/FL-147

Reopen the filter, the previously set settings will disappear.
no task

# Changes

This pull request primarily focuses on enhancing data filtering and pagination in the data grid component of the design systems package. The changes involve adding a new `isVisible` property to the `CustomFilter` component, using a more efficient way of handling filter changes, and improving the test cases for the `CustomFilter` component. The pull request also simplifies the way data is passed to the `useDataGrid` function in the `DataGridManualPaginationStory` component.

Enhancements to data filtering:

* [`apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx`](diffhunk://#diff-2de5b29ead6c7733af1a3b86e9cc9989b01fb6ce2881bec14aae18ae3ee67e82R22-R285): Added state variables for current filter and data, and an `onFilterChange` function to handle filter changes. This function filters the data based on the current filter state.

* [`packages/design-systems/src/components/composite/Datagrid/Datagrid.tsx`](diffhunk://#diff-df05b9030c0841de2c2cbf61a9e558434a1a78776196cd38d0b77e824095af7cL158-R166): Moved the `CustomFilter` component outside the `Table` component and added the `isVisible` property to control its visibility. [[1]](diffhunk://#diff-df05b9030c0841de2c2cbf61a9e558434a1a78776196cd38d0b77e824095af7cL158-R166) [[2]](diffhunk://#diff-df05b9030c0841de2c2cbf61a9e558434a1a78776196cd38d0b77e824095af7cL211-L220)

* [`packages/design-systems/src/components/composite/Datagrid/SearchFilter/CustomFilter.tsx`](diffhunk://#diff-ba24f0046bbb65e4b433f2c85c6fc8864b22af8057ca5adada1149daaab8e17cL26-R27): Added the `isVisible` property to the `CustomFilter` component and used the `usePrevious` hook to prevent unnecessary filter changes. [[1]](diffhunk://#diff-ba24f0046bbb65e4b433f2c85c6fc8864b22af8057ca5adada1149daaab8e17cL26-R27) [[2]](diffhunk://#diff-ba24f0046bbb65e4b433f2c85c6fc8864b22af8057ca5adada1149daaab8e17cR154-R163) [[3]](diffhunk://#diff-ba24f0046bbb65e4b433f2c85c6fc8864b22af8057ca5adada1149daaab8e17cR245-R249) [[4]](diffhunk://#diff-ba24f0046bbb65e4b433f2c85c6fc8864b22af8057ca5adada1149daaab8e17cR296-R303)

* [`packages/design-systems/src/components/composite/Datagrid/types.ts`](diffhunk://#diff-941d961c106632be8cee06d9e24ca67f13ba99f146bdee9eaa0fcbb782908917R71): Added the `isVisible` property to the `CustomFilterProps` type.

Improvements to test cases:

* [`packages/design-systems/src/components/composite/Datagrid/SearchFilter/CustomFilter.test.tsx`](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR120): Added the `isVisible` property to the `CustomFilter` component in all test cases. [[1]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR120) [[2]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR149) [[3]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR178) [[4]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR266) [[5]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR339) [[6]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR417) [[7]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR495) [[8]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR586) [[9]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR685) [[10]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR783) [[11]](diffhunk://#diff-7639ed2d86a4f46ee4ef496ff371b3ef6f083ec52dadc872fa2f28c7a774edfcR843)

Simplification of data passing:

* [`apps/storybook/src/stories/datagrid/DataGridManualPaginationStory.tsx`](diffhunk://#diff-137d0cbed8df6d3dc9bd10366cdca12fda403c6ef703eab1c2ef7126b07fc0ebL25-R25): Simplified the way data is passed to the `useDataGrid` function by using shorthand property notation.

### CODE REVIEW

File: apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx [hunk 1]
```diff
@@ -17,10 +19,270 @@ export type DataGridWithFilterStoryProps = {
 export const DataGridWithFilterStory = ({
   enableColumnFilters = false,
 }: DataGridWithFilterStoryProps) => {
+  const [currentFilter, setCurrentFilter] = useState({});
+  const [data, setData] = useState<Payment[]>(originData);
   const table = useDataGrid({
     data,
     columns,
     enableColumnFilters,
+    onFilterChange: (filter) => {
```
Comment (IzumiSy) [inline apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx:28]: ここから下のフィルタ機能のエミュレーション実装は切り出して他のstoryでもつかえるようにしてもらいたいです。

というのもDataGrid With Multilingual Supportのstoryでもフィルタのパネルが表示されるのですが、そちらでは動かない挙動になってしまっているので、このフィルタ挙動の実装を与えて動くようにしたいなと。

File: apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx [hunk 2]
```diff
@@ -17,10 +19,270 @@ export type DataGridWithFilterStoryProps = {
...
+      const isSameFilter =
+        JSON.stringify(filter) === JSON.stringify(currentFilter);
```
Comment (IzumiSy) [inline apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx:30]: datagrid側でフィルタの内容が同じであれば変更を通知しない実装しているのであれば、アプリケーション側でこの等価チェックは不要じゃないですかね?

Comment (murayama-r) [reply apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx:30]: 不要ですね。削除します

File: apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx (hunk 1 above)
Comment (murayama-r) [reply apps/storybook/src/stories/datagrid/DataGridWithFilterStory.tsx:28]: 承知しました。

**Review by IzumiSy — APPROVED** [review]

**Pull request merged**